docs/proposals/rescheduling.md

```diff
@@ -0,0 +1,496 @@
+  effects of decisions (e.g. whether an evicted pod will reschedule, will cause
+  a preemption when it reschedules, etc.), does the rescheduler execute multi-step plans
+  (e.g. evict two pods at the same time with the intent of moving one into the space
+  vacated by the other, or even more complex plans)
+
+Additional musings on the rescheduling design space can be found [here](rescheduler.md).
+
+## Design proposal
+
+The key mechanisms and components of the proposed design are priority, preemption,
+disruption budgets, the `/evict` subresource, and the rescheduler.
+
+### Priority
+
+#### Motivation
+
+[Quality of Service levels](resource-qos.md) allow Kubernetes to safely overcommit nodes.
+Just as it is useful to overcommit nodes to increase node-level utilization, it is useful
+to overcommit clusters to increase cluster-level utilization.  Scheduling priority (which
+we abbreviate as *priority*, in combination with disruption budgets (described in the
+next section), allows Kubernetes to safely overcommit clusters much as QoS levels allow
+it to safely overcommit nodes.
+
+Today, cluster sharing among users, workload types, etc. is regulated via the
+[quota](../admin/resourcequota/README.md) mechanism. When allocating quota, a cluster
+administrator has two choices: (1) the sum of the quotas is less than or equal to the
+capacity of the cluster, or (2) the sum of the quotas is greater than the capacity of the
+cluster (that is, the cluster is overcommitted).  (1) is likely to lead to cluster
+under-utilization, while (2) is unsafe in the sense that someone's pods may go pending
+indefinitely even though they are still within their quota. Priority makes cluster
+overcommitment (i.e. case (2)) safe by allowing users and/or administrators to identify
+which pods should be allowed to run, and which should go pending, when demand for cluster
+resources exceeds supply to due to cluster overcommitment.
+
+Priority is also useful in some special-cause scenarios, such as ensuring that system
+DaemonSets can always schedule and reschedule onto every node where they want to run
+(assuming they are given the highest priority), e.g. see #21767.
+
+#### Specifying priorities
+
+We propose to add a required `Priority` field to `PodSpec`. Its value type is string, and
+the cluster administrator defines a total ordering on these strings (for example
+`Critical`, `Normal`, `Preemptible`). We choose string instead of integer so that it is
+easy for an administrator to add new priority levels in between existing levels, to
+encourage thinking about priority in terms of user intent and avoid magic numbers, and to
+make the internal implementation more flexible.
+
+When a scheduler is scheduling a new pod P and cannot find any node that meets all of P's
+scheduling predicates, it is allowed to evict ("preempt") one or more pods that are at
+the same or lower priority than P (subject to disruption budgets, see next section) from
```

Comment:
> I guess the assumption is that the evicted pod might be schedulable on some other node. In the worst case we break even (modulo rescheduling costs), while in the best case we have an additional pod scheduled.

Correct.

> One can imagine a pretty bad situation where the above worst-case behavoir cascades, evicting a large number of equal-priority pods in succession, resulting in high rescheduling costs, and no net benefit.

As @mqliang says, preemptions are rate-limited by disruption budgets, so you can't get preemption cascades/preemption storms.

> Would it be worth trying to short-circuit this sort of thing (by e.g. only pre-empting an equal-priority pod if it is immediately schedulable onto another node without evicting anything, or only evicting lower-priority pods?

What you're describing (only evict if we're pretty sure the victim can reschedule somewhere) is something we might eventually want to do for the other rescheduling use cases, and I agree it could be useful for the same-priority-preemption use case as well. I don't think it's necessary for a first version of the system, though.

>  I have the same concern of why we could allow equal priority preemption.

The reason to allow equal-priority preemption is to "repack" pods to make room for a pending pod. (For example, defragment resources, or move a pod off of a machine with some label that is required by the pending pod but not required by the already-running pod.) Doing this at scheduling time on-demand rather than speculatively in the rescheduler minimizes the amount of unnecessary evictions.